### ISSUE TITLE
Unusual Inference Time Behavior with GPU Usage

### ISSUE THREAD

**Issue by amit154154** [facebookresearch/dinov2#250]
Hello everyone,

I've encountered an unusual issue during inference that I need some help with. I've tried this on three different computers, including Google Colab, and the problem persists across all of them.

I load a model using the following code:

`image_encoder = torch.hub.load('facebookresearch/dinov2', 'dinov2_vitb14').cuda().eval()`

Then, I run inference as shown below:

```
import time
t = time.time()
with torch.no_grad():
    mask_features = image_encoder(torch.randn(20,3,224,224).cuda())
time.time() - t
```
Surprisingly, the above code executes in the expected 0.05 seconds. However, when I add any GPU-related operation after the inference, such as creating a new tensor, the execution time increases significantly, as demonstrated in the following code snippet:
```
import time
t = time.time()
with torch.no_grad():
    mask_features = image_encoder(torch.randn(20,3,224,224).cuda())
x = torch.randn(1).cuda()
time.time() - t
```
The second code snippet takes 0.39 seconds to run, which is much longer than expected.

I find this behavior quite perplexing and would greatly appreciate any insights or assistance in understanding and resolving this issue.

Thank you in advance! (:

**Comment by qasfb** [comment]
That's because gpu calls are asynchronous so python gets control immediately after calling model inference.
However, in the second snippet you do a .cuda() call, which copies from CPU to GPU, which waits for all GPU operations to complete before proceeding.

If you want to have a proper timing, you should add torch.cuda.synchronize() after the model inference call. If you want to measure timing properly you should do it in a loop, and measure an average after a few warmup calls.

**Comment by WeijianZhang123** [comment]
I have also encountered the same problem on different GPUs and am currently unable to solve it.

![Image](https://github.com/user-attachments/assets/c9abc6ad-cfb3-4a1a-9647-ef113c0766e7)
